x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ReplaceDataStreamBackingIndexStep.java

```diff
@@ -86,14 +86,18 @@ public ClusterState performAction(Index index, ClusterState clusterState) {
             throw new IllegalStateException(errorMessage);
         }
 
-        assert dataStream.getWriteIndex() != null : dataStream.getName() + " has no write index";
-        if (dataStream.getWriteIndex().equals(index)) {
+        Index dataStreamWriteIndex = dataStream.isFailureStoreIndex(originalIndex)
+            ? dataStream.getFailureStoreWriteIndex()
+            : dataStream.getWriteIndex();
+        assert dataStreamWriteIndex != null : dataStream.getName() + " has no write index";
+        if (dataStreamWriteIndex.equals(index)) {
             String errorMessage = String.format(
                 Locale.ROOT,
-                "index [%s] is the write index for data stream [%s], pausing "
+                "index [%s] is the%s write index for data stream [%s], pausing "
                     + "ILM execution of lifecycle [%s] until this index is no longer the write index for the data stream via manual or "
                     + "automated rollover",
                 originalIndex,
+                dataStream.isFailureStoreIndex(originalIndex) ? " failure store" : "",
```

Comment:
Tiny nit: I think it makes sense to align this code with the other checks you're doing (i.e. the `targetWriteIndex` and `isFailureStoreIndex` variables), that makes them slightly more readable IMO.

Reply:
Sorry I should have phrased this differently I think. Since you went ahead with my suggestion in [this](https://github.com/elastic/elasticsearch/pull/108741#discussion_r1606501087) comment, I think this comment still applies -- that is, the handful of checks you're adding in this PR should be aligned with each other.

Reply:
Updated in https://github.com/elastic/elasticsearch/pull/108741/commits/4e62a80b987fa361ee14a425342cfb12426ae081